### ISSUE TITLE
Improve access to important branches

### ISSUE THREAD

**Issue by mstv** [gitextensions/gitextensions#8741]
## Feature description

(Not only) the left panel can be overcrowded with foreign remote branches, which are of little interest for me - especially in long running projects with many contributors. (I know branch naming conventions could help but not fully.)
I want to see an overview of important branches (filtered by regex), e.g. "main", "develop", "release/xx.yy".
Either by tinkering the sort order or rather by showing them again under an additional node. E.g. new nodes "Important" and "All" could be inserted under the "Remotes" node.

This filter could also be applied to "Contained in branches:" of the CommitInfo panel.

Thoughts?

## Environment

- Git Extensions 33.33.33
- Build f96fa18202704df0606d41ff2f5c24e7f7ccc107
- Git 2.27.0.windows.1 (recommended: 2.28.0 or later)
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

**Comment by gerhardol** [comment]
The same pattern should be presented first in RevGrid (and CommitInfo as stated) too. (There is some hardoding for branches starting with master in CommitInfo.cs right now.)
At work, these branches are "protected". Maybe "Prioritized" is a better concept than "Important"?

Slightly relate: I want to relate the local<->remote branch in a simpler way. Ideally selecting both by Alt-click, but a menu item is easier to handle. (Separate discussion.)

**Comment by ivangrek** [comment]
Two features.

1) Filter nodes in the left panel by input value.
2) Be able to pin nodes and not sort about general order.

**Comment by gerhardol** [comment]
>     1. Filter nodes in the left panel by input value.

Can you explain?

>     2. Be able to pin nodes and not sort about general order.

I would rather add a pattern to not have to pin important branches explicitly (and to add release/2021-01, release/2021-02 etc).
But these 'important' branches could be displayed above the 'sorted' branches.

**Comment by ivangrek** [comment]
> Can you explain?

![image](https://user-images.githubusercontent.com/3169265/104442810-45d96180-559e-11eb-8034-b4cb2bc2994d.png)

> I would rather add a pattern to not have to pin important branches explicitly (and to add release/2021-01, release/2021-02 etc).
But these 'important' branches could be displayed above the 'sorted' branches.

Also filter by date and then pin the filtered ones.

And one more feature.

3. Ability to create a folder with the ability to set filters and show filtered nodes in the folder.

**Comment by RussKie** [comment]
> the left panel can be overcrowded with foreign remote branches

Could you please clarify this point?

I tend to have a lot of remotes, but I toggle them on/off as necessary. And I don't generally don't have a need to keep the remotes expanded.
If a remote has too many branches (especially the trunk) then it is more a business/team discipline issue, and perhaps regular pruning and forks should be considered.  
I've been down this road several time, and such issues are rarely fixable by tools only. Besides there are concrete financial benefits of clearing old branches - less stress on your git hoster (on-prem or cloud) and less space required.

For completeness, the "Branches" node generally represents the user's private space - here are user's branches, and it is up to the user to organise his/hers personal space as required. Too many branches - prune them, or come up with a naming convention that hides them (e.g. use folders).

. . .

Filtering is a sore point in the app. That was partially on of many reasons why I want to split the toolbar and show/hide individual toolbars - the current filter toolbar is close to useless.

In the left panel specifically I [proposed to move filtered out branches into "Hidden" node](https://github.com/gitextensions/gitextensions/pull/8414#issuecomment-683248073), but the proposal was rejected (stating the fact, no hard feelings).

Before we start heavily invest into filtering in the left panel I believe we'd gain significantly more if we look at filtering holistically, and improve the filtering overall. By filtering branches in the graph, we'll automatically benefit from filtering in the left panel.

For instance, dotnet/winforms repo has several release branches that are active, but I have several clones due to build nuances of each branch. Because the branches are live, there is a bit of a noise, and when I work on a release/5.0 branch I don't want to see the master or release/3.1 branches, but all branches that belong to the current branch. 

Be able to filter out branches in this fashion would deliver a significantly greater value to everyone.

**Comment by gerhardol** [comment]
> If a remote has too many branches (especially the trunk) then it is more a business/team discipline issue, and perhaps regular pruning and forks should be considered.

I do not disagree that the workflow should be changed, but it still happens. This discussion is on the way to improving.

> Before we start heavily invest into filtering in the left panel I believe we'd gain significantly more if we look at filtering holistically, and improve the filtering overall. By filtering branches in the graph, we'll automatically benefit from filtering in the left panel.

Probably, I have no great idea myself

> For instance, dotnet/winforms repo has several release branches that are active, but I have several clones due to build nuances of each branch. Because the branches are live, there is a bit of a noise, and when I work on a release/5.0 branch I don't want to see the master or release/3.1 branches, but all branches that belong to the current branch.

You should use worktrees...

**Comment by vbjay** [comment]
If you name your branches with / GE will create folders in the tree.

Release/v2.2/pre


Release/2.2/fin

**Comment by mstv** [comment]
> If a remote has too many branches (especially the trunk) then it is more a business/team discipline issue, and perhaps regular pruning and forks should be considered. ...

As for me, you are preaching to the converted. Unfortunately, I am not in the position to push / enforce this - expressing it kindly - in a very heterogeneous world-wide team, especially in the current project phase.

As I stated: I want to see an overview of important branches (filtered by regex), e.g. "main", "develop", "release/xx.yy".
E.g. older releases are not of interest anymore but come first when sorted A-Z.

I want to see the unfiltered graph (most of the time). But I want to quickly access the important/prioritized branches (using the left panel).

> Be able to filter out branches in this fashion would deliver a significantly greater value to everyone.

Perhaps you and your colleagues would benefit from _multiple_ "folder nodes" with different regex.
These nodes could also be used to set the filter for the RevGrid or globally, e.g. on Alt-Click and context menu.

**Comment by RussKie** [comment]
> You should use worktrees...

I should try them one day :)



> Perhaps you and your colleagues would benefit from _multiple_ "folder nodes" with different regex.

This is an interesting idea 👍

**Comment by pmiossec** [comment]
> > You should use worktrees...
> 
> I should try them one day :)

Yes, that's not difficult to understand. 15 min investment to get it and it could really help...

And that way you will be able to improve worktree support in GE 🤣

**Comment by Piedone** [comment]
I just wanted to open a similar for a similar case: Lot of remote branches in the repo (and lot of local ones too since I checked them out at one point) sot he Branches and Remotes/origin sections on the left panel are growing continuously.

My suggestion would be to have an option to only show recent (recently used) branches, much like how [GitHub's branch list](https://github.com/gitextensions/gitextensions/branches) does (though this seems to be a bit broken now with apparently no active branches in this repo :)). That way, you don't have to configure any filtering and the important branches (i.e. the ones that are frequently used) would show up there. I'd use a threshold (perhaps also configurable?) as "recent", with a default of maybe 10 days.

**Comment by Piedone** [comment]
Additionally, the branch text box in the toolbar could be bigger (or auto-grow) to accommodate for longer branch names (it's really small now), and auto-complete for remote branches (i.e. the ones prefixed with e.g. `origin/`) too.

**Comment by gerhardol** [comment]
In the context menu (inaddition to Settings as seen above)
![image](https://user-images.githubusercontent.com/6248932/162962435-c6209b13-6f69-4ec5-b95f-4768cfd39227.png)

**Comment by gerhardol** [comment]
Draft implementation in fc45721ab92b80cdcf624dd59838f2a00da5f911
![image](https://user-images.githubusercontent.com/6248932/197406250-12b57d02-d42a-4935-a480-ee12099ddf0e.png)
![image](https://user-images.githubusercontent.com/6248932/197406277-a5407737-6970-4b4d-b6df-2001166cad0a.png)

Just added to a settings dialog, no translation update. Could be added to the context menu or Appearance.
![image](https://user-images.githubusercontent.com/6248932/197406363-3190fc91-e5fe-43ff-b365-90cad22f5ecf.png)

**Comment by Piedone** [comment]
This is just about filtering branches, as opposed to ordering (like I mentioned [here](https://github.com/gitextensions/gitextensions/issues/8741#issuecomment-765709566), right?

**Comment by gerhardol** [comment]
> This is just about filtering branches, as opposed to ordering (like I mentioned [here](https://github.com/gitextensions/gitextensions/issues/8741#issuecomment-765709566), right?

The poc I added is for the intent for this PR: Improve access, no filtering.
Only in side panel.

You can configure sorting to show most recent used first already, this setting is global.
No option to limit the number of branches visible in sidepanel, you probably want all branches there. Some limit in other menus.

**Comment by mstv** [comment]
Applied to CommitInfo "Contained in branches": 45841b472a4ec4297d4538c739b47c2d1604bd7d

**Comment by gerhardol** [comment]
> Applied to CommitInfo "Contained in branches": [45841b4](https://github.com/gitextensions/gitextensions/commit/45841b472a4ec4297d4538c739b47c2d1604bd7d)

So add to Aooearance. Just add a link to settings in the menu.

- FormPush: unique default remote for new branches

How do you relate these features?
(I assume this was a todo just want to be sure I did not miss anything).

- remove ROT from name

Yes

> - precompile regexes
- but use them directly without restart

Not necessarily better. In sidepanel, the first time I neasured the time required to 1 ms wusing stopwatch. Next run was in us. The last 15 regex are cached.

**Comment by mstv** [comment]
> > Applied to CommitInfo "Contained in branches": [45841b4](https://github.com/gitextensions/gitextensions/commit/45841b472a4ec4297d4538c739b47c2d1604bd7d)
> 
> So add to Appearance.

Yes, to the GroupBox `General`.

> Just add a link to settings in the menu.

Which menu do you mean? Will be OK anyway.

>>  * FormPush: unique default remote for new branches 
> 
> How do you relate these features? (I assume this was a todo just want to be sure I did not miss anything).

Yes, it is a separate issue but related to sorting. I have removed the `z_` prefix from foreign remotes- and your's has become the default push target...

> > * precompile regexes
> 
>     * but use them directly without restart
> 
> 
> Not necessarily better. In sidepanel, the first time I measured the time required to 1 ms using stopwatch. Next run was in us. The last 15 regex are cached.

You are the expert.

**Comment by gerhardol** [comment]
> > So add to Appearance.
> 
> Yes, to the GroupBox `General`.
> 
> > Just add a link to settings in the menu.
> 
> Which menu do you mean? Will be OK anyway.

No setting, open settings in addition to the existing items

![image](https://user-images.githubusercontent.com/6248932/201485231-75a9efb8-8bbf-4103-94ad-c5cc3407b91b.png)

> > > * FormPush: unique default remote for new branches
> > 
> > 
> > How do you relate these features? (I assume this was a todo just want to be sure I did not miss anything).
> 
> Yes, it is a separate issue but related to sorting. I have removed the `z_` prefix from foreign remotes- and your's has become the default push target...

Well, you opened the issue and may redefine the scope still to include remotes too. Would affect the disabled remotes too in FormRemote in addition to FormPush.

But that raises another question: You probably want origin to be first, always, then upsteam. That would be priority OrderBy so the expression `^(origin|upstream)$` could be written as `^(origin)$;^(upstream)$` so it can be split in two. Will users understand that?

**Comment by mstv** [comment]
> That would be priority OrderBy so the expression ^(origin|upstream)$ could be written as ^(origin)$;^(upstream)$ so it can be split in two. Will users understand that?

Yes, they would. But this syntax would need much more explanation than a regex.
Default push target - perhaps in dependency on the branch name - is a separate feature.

I have moved `^(   )$` to the code. There is no need to bother users with it. If they (including you) want to match branches starting with a string, just append `.*` instead of omitting `$`. This is clearer.

**Comment by RussKie** [comment]
> But that raises another question: You probably want origin to be first, always, then upsteam. That would be priority OrderBy so the expression `^(origin|upstream)$` could be written as `^(origin)$;^(upstream)$` so it can be split in two. Will users understand that?

I use "remote" instead of "upstream" (i.e., the main trunk), and "origin" for my forks... Perhaps this list should be configurable in some way.

**Comment by RussKie** [comment]
> Could be added to the context menu or Appearance.

I wouldn't do this in a context menu, for a multitude of reasons including UX and accessibility. "Appearance" tab sounds like a good idea, though I kind of feel we need to add a subtab "Sorting" to which we'll move all sort-related configurations:
![image](https://user-images.githubusercontent.com/4403806/201518720-f762ac83-8ef1-467a-8541-04a873583993.png)

**Comment by gerhardol** [comment]
> > Could be added to the context menu or Appearance.
> 
> I wouldn't do this in a context menu, for a multitude of reasons including UX and accessibility. "Appearance" tab sounds like a good idea, though I kind of feel we need to add a subtab "Sorting" to which we'll move all sort-related configurations

If this was only for the sidepanel, it would have been natural to have it together with the other options. Probably easier to implement priorities everywhere then to fix the menu.

If moving out the current four and two new to a separate tab, should the six lines for Author images be separated too?
Will keep the group-box, even if there is only one group right now.

**Comment by RussKie** [comment]
I'd move the sorting only, keep everything else in tact. I just don't think all sort-related configs will fit on the page at its smallest dimensions.

**Comment by gerhardol** [comment]
> > That would be priority OrderBy so the expression ^(origin|upstream)$ could be written as ^(origin)$;^(upstream)$ so it can be split in two. Will users understand that?
> 
> Yes, they would. But this syntax would need much more explanation than a regex. Default push target - perhaps in dependency on the branch name - is a separate feature.

The ordering is getting weird with multiple levels, has to be something like the below.
Will obviously be a common method

```patch
diff --git a/GitUI/BranchTreePanel/LocalBranchTree.cs b/GitUI/BranchTreePanel/LocalBranchTree.cs
index c6ff8..b1a4c 100644
--- a/GitUI/BranchTreePanel/LocalBranchTree.cs
+++ b/GitUI/BranchTreePanel/LocalBranchTree.cs
@@ -52,13 +52,33 @@ protected override Nodes FillTree(IReadOnlyList<IGitRef> branches, CancellationT
 
             Nodes nodes = new(this);
             var aheadBehindData = _aheadBehindDataProvider?.GetData();
-
-            // Sort prio branches first (if set) with the compile cache (no need to instanceiate)
-            bool isBranchRegex = !string.IsNullOrWhiteSpace(AppSettings.RepoObjectsTreePrioBranchNames);
             string currentBranch = Module.GetSelectedBranch();
             Dictionary<string, BaseRevisionNode> pathToNode = new();
-            foreach (IGitRef branch in branches.OrderBy(node =>
-                    isBranchRegex && !Regex.IsMatch(node.LocalName, $"^({AppSettings.RepoObjectsTreePrioBranchNames})$", RegexOptions.ExplicitCapture)))
+
+            // Sort prio branches first (if set) with the compile cache (no need to instanceiate)
+            string[] prios = string.IsNullOrWhiteSpace(AppSettings.RepoObjectsTreePrioBranchNames)
+                ? Array.Empty<string>()
+                : AppSettings.RepoObjectsTreePrioBranchNames.Split(';', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries);
+            List<IGitRef> orderedBranches;
+            if (prios.Length == 0)
+            {
+                orderedBranches = (List<IGitRef>)branches;
+            }
+            else
+            {
+                IOrderedEnumerable<IGitRef> ordering = branches
+                .OrderBy(node =>
+                    !Regex.IsMatch(node.LocalName, $"^({prios[0]})$", RegexOptions.ExplicitCapture));
+                foreach (var s in prios.Skip(1))
+                {
+                    ordering = ordering.ThenBy(node =>
+                        !Regex.IsMatch(node.LocalName, $"^({s})$", RegexOptions.ExplicitCapture));
+                }
+
+                orderedBranches = ordering.ToList();
+            }
+
+            foreach (IGitRef branch in orderedBranches)
             {
                 token.ThrowIfCancellationRequested();
```

**Comment by gerhardol** [comment]
@mstv How to proceed here?
Should we create a PR with the changes in here for now?

* Appearance settings page is already overcrowded,
* Multi priority is hard to implement in a general way, could be OK for sidepanel though, but then the regex will have to be modified for commit info.
* This should be a general setting, applying to all branch and remote lists, which is tricky.

**Comment by mstv** [comment]
> Should we create a PR with the changes in here for now?

Yes, please, including afb4688f000b478fadeec95192ff1420289902e9. (I have not updated the TODO list.)

(I think the default push-to-repo should be configured in the Repo Manager, i.e. it is orthogonal to this feature.)

> Multi priority is hard to implement in a general way, could be OK for sidepanel though, but then the regex will have to be modified for commit info.
This should be a general setting, applying to all branch and remote lists, which is tricky.

I do not have a use case myself. @RussKie?

> Appearance settings page is already overcrowded,

The space should suffice for these two boxes. If you are willing to or most likely for "multi priority", you could go for https://github.com/gitextensions/gitextensions/issues/8741#issuecomment-1312704637.

**Comment by RussKie** [comment]
> > Multi priority is hard to implement in a general way, could be OK for sidepanel though, but then the regex will have to be modified for commit info.
> > This should be a general setting, applying to all branch and remote lists, which is tricky.
> 
> I do not have a use case myself. @RussKie?

TBH, I'm getting by having the existing sort options. 
The only situation that comes to mind is when I work in repos without forks (i.e., everyone is working off the main trunk), and I find myself hunting for my branches. I suppose, prioritising my own branches above the rest could help, but to me that still feels like a relatively niche feature.

**Comment by gerhardol** [comment]
draft with priorities in branch feature/i8741-important-branches

> The only situation that comes to mind is when I work in repos without forks (i.e., everyone is working off the main trunk), and I find myself hunting for my branches. I suppose, prioritising my own branches above the rest could help, but to me that still feels like a relatively niche feature.

Exactly what I work with, similar to several other colleagues in other companies.
Private forks are disabled.
First you have to convince IT that private forks will not have an effect on storage, then you have to teach all central repo users what the difference is between the private fork and the master fork.
There is one minor benefit, it is easier to see if someone is working on an issue.

**Comment by RussKie** [comment]
Hear hear :)
